### PR TITLE
flowinfra: fix gRPC stream leak

### PR DESCRIPTION
**Pull request by andreimatei**
The outbox wasn't always properly handling the gRPC stream from the consumer. As per the gRPC contract hidden in [1], one can't simply walk away from a stream; it needs to make sure it's closed in one of a number of ways, otherwise gRPC leaks goroutines.
Taking that contract to heart, our stream tracing also leaks spans if the stream is not disposed of properly. We were not hitting this part, though, because the outbox uses a dubious context.TODO() when establishing the stream - so our tracing was inhibited. But I was doing what was supposed to be an innocuous change to create a client-side span for such a stream, only to hit it.

[1] https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream

Release note: None
Epic: None

### CODE REVIEW

**Comment by andreimatei** [comment]
Yahor, I don't know if the drain I've done is the right thing to do. I'll let you study whether we should be associating a context with the stream and canceling it instead, or what.

Relatedly, do you mind if I ask you to look at your TODO [here](https://github.com/cockroachdb/cockroach/blob/f95b04688a1a4e94e6d7bd87b3d20b159086783f/pkg/sql/flowinfra/outbox.go#L237-L240) about the `context.TODO` used for this stream? Perhaps that will show us the right answer.

**Review by yuzefovich — APPROVED** [review]
Addressing that TODO has been on my plate for a while now (latest attempt was #90864 which ran into some CI failures that I never really examined), so I'll take a closer look soonish. In the mean time, if this is blocking you, :lgtm: to merge this and I'll tweak it later if I discover another solution.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93840#-:-NJfUj5jAaSxoxXxQJL8:bv2xfk4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

**Comment by yuzefovich** [comment]
Actually, let's hold off on merging this given that #93542 still has some flakes and I'm making progress in #90864.

**Comment by yuzefovich** [comment]
Alright, CI seems to be ok with #90864, I'm curious if you rebase your work on top of that change and see if things improve.

**Comment by andreimatei** [comment]
For the purposes of https://github.com/cockroachdb/cockroach/pull/93542, there's at least one more stream leak in a random test that's causing a failure. I has seen it when looking at one of the failing tests, but I forgot which one.
Would you mind patching [`7213a0d` (#93542)](https://github.com/cockroachdb/cockroach/pull/93542/commits/7213a0dae3eccbde66d72848aabc36efcbe0b15f) on your https://github.com/cockroachdb/cockroach/pull/90864 and checking how the test failures look compared to the current form of #93542 [i.e. here](https://teamcity.cockroachdb.com/viewLog.html?buildId=7999883&buildTypeId=Cockroach_BazelEssentialCi) ?
Btw, there's a failure for `TestStreamConnectionTimeout` which seems related, even with this patch.

**Pull request closed without merging**
